Library/Homebrew/utils/pypi.rb

```diff
@@ -0,0 +1,113 @@
+    pypi_name = if formula.stable.url.start_with?("https://files.pythonhosted.org/packages/")
+      File.basename(formula.stable.url).match(/^(.+)-[a-z0-9.]+$/)[1]
+    else
+      formula.name
+    end
+
+    version ||= formula.version
+
+    # See if package is a PyPI package
+    if get_pypi_info(pypi_name, version).nil?
+      odie "\"#{pypi_name}\" at version #{version} is not available on PyPI." unless ignore_non_pypi_packages
+      return
+    end
+
+    # Ensure there are no non-pypi resources when not running with --print-only
+    non_pypi_resources = formula.resources.reject do |resource|
+      resource.url.start_with? "https://files.pythonhosted.org/packages/"
+    end
+    if !non_pypi_resources.empty? && !print_only
+      odie "\"#{formula.name}\" contains non-PyPI resources. Please update the resources manually."
+    end
+
+    # Get list of required packages
+    ohai "Retrieving PyPI dependencies for \"#{pypi_name}==#{version}\"" unless print_only || silent
+    pipgrip_output = Utils.popen_read Formula["pipgrip"].bin/"pipgrip", "--json", "#{pypi_name}==#{version}"
+    unless $CHILD_STATUS.success?
+      odie <<~EOS
+        Unable to determine dependencies for \"#{pypi_name}\" because of a failure when running
+        `pipgrip --json #{pypi_name}==#{version}`. Please update the resources for \"#{formula.name}\" manually.
+      EOS
+    end
+
+    packages = JSON.parse(pipgrip_output).sort.to_h
+
+    # Remove extra packages that may be included in pipgrip output
+    exclude_list = %w[argparse pip setuptools wsgiref] + [pypi_name]
+    packages.delete_if do |package|
+      exclude_list.include? package
+    end
+
+    # Create resource blocks
+    new_resource_blocks = ""
+    packages.each do |package, package_version|
+      name, url, checksum = get_pypi_info package, package_version
+      # Fail if unable to find name, url or checksum for any resource
+      if name.nil?
+        odie "Unable to resolve some dependencies. Please update the resources for \"#{formula.name}\" manually."
+      elsif url.nil? || checksum.nil?
+        odie <<~EOS
+          Unable to find the URL and/or sha256 for the \"#{name}\" resource.
+          Please update the resources for \"#{formula.name}\" manually.
+        EOS
+      end
+
+      # Indent and append each resource block
+      new_resource_blocks += "  " + <<~EOS.chomp.gsub("\n", "\n  ") + "\n\n"
+        resource "#{name}" do
+          url "#{url}"
+          sha256 "#{checksum}"
+        end
+      EOS
+    end
+
+    # Print resource lines only if --print-only is passed
+    puts new_resource_blocks.chomp if print_only
+    return if print_only
+
+    # Check whether resources already exist (excluding homebrew-virtualenv)
+    if formula.resources.empty? ||
+       (formula.resources.length == 1 && formula.resources.first.name == "homebrew-virtualenv")
+      # Place resources above install method
+      inreplace_regex = /  def install/
```

Comment:
Don't let it block this PR but this seems like something `rubocop`'s `parser` library would be able to handle better/more reliably than regex.

Reply:
I agree. I'll have to look into how to handle this (I'm not sure how to use the parse in the non-rubocop areas). Are there any other commands/places that use the parser that I can look at for reference?

Reply:
I don't think anything has used this before 😭. You could perhaps hack it into just calling RuboCop at runtime or dig into the `parser` gem? I think it would be something useful for `brew bottle`, too, if you got it working. As said: not blocking.